File: 3299-find-the-maximum-number-of-elements-in-subset/find-the-maximum-number-of-elements-in-subset.java
class Solution {
    public int maximumLength(int[] nums) {
        Map<Long, Integer> cnt = new HashMap<>();
        for (int num : nums) {
            cnt.merge((long) num, 1, Integer::sum);
        }

        int oneCnt = cnt.getOrDefault(1L, 0);
        int maxLength = oneCnt - 1 + (oneCnt & 1);
        cnt.remove(1L);

        for (long num : cnt.keySet()) {
            int curLength = 0;
            while (cnt.containsKey(num)) {
                if (cnt.get(num) == 1) {
                    curLength += 1;
                    break;
                }
                curLength += 2;
                num *= num;
            }
            maxLength = Math.max(maxLength, curLength - 1 + (curLength & 1));
        }

        return maxLength;
    }
}

File: 3299-find-the-maximum-number-of-elements-in-subset/find-the-maximum-number-of-elements-in-subset.py
class Solution:
    def maximumLength(self, nums: List[int]) -> int:
        dp = dict()         # start -> maxLength
        freqs = collections.Counter(nums)

        # handle edge case
        maxOnes = freqs[1] if freqs[1] & 1 == 1 else freqs[1] - 1
        freqs.pop(1, None)
        dp[1] = maxOnes

        # print (freqs)
        for start, freq in freqs.items():
            # print(start, freq)
            if freq == 1:
                dp[start] = 1
                continue

            length = 1
            nxt = start * start
            while nxt in freqs:
                length += 1
                if freqs[nxt] == 1:
                    break
                nxt *= nxt

            dp[start] = 2 * length - 1

        # print(dp)
        return max(dp.values())